File: src/decoders/news.ts
import type { News, NewsCategory, SessionHandle } from "~/models";
import { decodeNewsInformation } from "./news-information";
import { decodeNewsCategory } from "./news-category";
import { decodeNewsSurvey } from "./news-survey";

export const decodeNews = (news: any, session: SessionHandle): News => {
  const categories: NewsCategory[] = news.listeCategories.V.map(decodeNewsCategory);

  return {
    categories,
    items: news.listeModesAff[0].listeActualites.V.map((item: any) => {
      let decoder: typeof decodeNewsInformation | typeof decodeNewsSurvey;

      if (item.estInformation) decoder = decodeNewsInformation;
      else if (item.estSondage) decoder = decodeNewsSurvey;
      else throw new Error("Unknown news type");

      return decoder(item, session, categories);
    })
  };
};

File: library/src/commonMain/kotlin/decoders/news.kt
package decoders

import kotlinx.serialization.json.*
import models.News
import models.SessionInformation
import models.errors.UnreachableError

fun decodeNews (news: JsonObject, sessionInformation: SessionInformation): News {
    val categories = news["listeCategories"]!!.jsonObject["V"]!!.jsonArray.map { decodeNewsCategory(it.jsonObject) }

    return News(
        categories = categories,
        items = news["listeModesAff"]!!.jsonArray[0].jsonObject["listeActualites"]!!.jsonObject["V"]!!.jsonArray.map {
            val decoder: (obj: JsonObject) -> Any

            if (it.jsonObject["estInformation"]!!.jsonPrimitive.boolean) decoder = {obj -> decodeNewsInformation(obj, sessionInformation) }
            else if (it.jsonObject["estSondage"]!!.jsonPrimitive.boolean) decoder = {obj -> decodeNewsSurvey(obj, sessionInformation, categories) }
            else throw UnreachableError("decodeNews")

            decodeNewsItem(it.jsonObject, categories, decoder)
        }
    )
}